File: auto_verbs/heavy-anti-bot-sites/udemy_com/udemy_search.py
"""
Udemy – Search "Python programming" → sort Highest Rated → extract top 5 courses.
Pure Playwright – no AI.
"""
import re, os, sys, traceback, shutil, tempfile, threading
from playwright.sync_api import Playwright, sync_playwright

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, launch_chrome, wait_for_cdp_ws

from dataclasses import dataclass




@dataclass(frozen=True)
class UdemySearchRequest:
    query: str = "Python programming"
    max_results: int = 5


@dataclass(frozen=True)
class UdemyCourse:
    title: str
    instructor: str
    rating: str
    price: str


@dataclass(frozen=True)
class UdemySearchResult:
    query: str
    courses: list


def search_udemy_courses(playwright, request: UdemySearchRequest) -> UdemySearchResult:
    port = get_free_port()
    profile_dir = tempfile.mkdtemp(prefix="udemy_")
    chrome_proc = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)
    browser = playwright.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()
    courses = []

    # Watchdog: force-kill everything after 90s so the script never hangs
    def _watchdog():
        print("\n⏱️  WATCHDOG: 90s timeout — force-killing Chrome...")
        try:
            chrome_proc.kill()
        except Exception:
            pass
        os._exit(1)

    timer = threading.Timer(90, _watchdog)
    timer.daemon = True
    timer.start()

    try:
        # STEP 1: Navigate directly to search results URL (avoids homepage bot detection)
        import urllib.parse
        q = urllib.parse.quote_plus(request.query)
        search_url = f"https://www.udemy.com/courses/search/?q={q}"
        print(f"STEP 1: Navigate to search results for '{request.query}'...")
        print(f"   URL: {search_url}")
        page.goto(search_url, wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(4000)
        print(f"   ✅ Landed on: {page.url}")

        # Dismiss cookie consent wall FIRST — it blocks page content
        print("   Dismissing cookie/consent banners...")
        for sel in [
            "button:has-text('Accept All')",
            "button:has-text('Accept all')",
            "button:has-text('Accept')",
            "button:has-text('Agree')",
            "button[data-purpose='accept-cookie']",
            "button:has-text('Dismiss')",
            "[aria-label='Close']",
            "#onetrust-accept-btn-handler",
            "button.cookie-consent--accept",
        ]:
            try:
                loc = page.locator(sel).first
                if loc.is_visible(timeout=1000):
                    loc.click()
                    print(f"   Dismissed: {sel}")
                    page.wait_for_timeout(1500)
                    break
            except Exception:
                pass

        page.wait_for_timeout(3000)

        # Scroll to load more courses
        for _ in range(8):
            page.evaluate("window.scrollBy(0, 600)")
            page.wait_for_timeout(800)

        print("STEP 3: Extract courses...")
        print(f"   Current URL: {page.url}")
        
        # Debug: check what's on the page
        h3_count = page.locator("h3").count()
        div_count = page.locator("div").count()
        course_card_count = page.locator("[data-purpose*='course'], [class*='course-card'], [class*='CourseCard']").count()
        print(f"   Page has: {div_count} divs, {h3_count} h3s, {course_card_count} course cards")
        if h3_count > 0:
            sample = page.locator("h3").first.text_content() or ""
            print(f"   First h3: '{sample.strip()[:80]}'")
        # Check page title to detect bot block
        title = page.title()
        print(f"   Page title: '{title[:80]}'")

        # Strategy 1: parse body text for course patterns
        # Pattern: Title line → description with "Rating: X.X out of 5" →
        #          "Instructor:" → name → "Rating:" → number → "(count)" → price
        seen = set()
        body = page.inner_text("body")
        lines = [l.strip() for l in body.splitlines() if l.strip()]

        i = 0
        while i < len(lines) and len(courses) < request.max_results:
            # Look for "Instructor:" marker
            if lines[i] == "Instructor:":
                instructor = lines[i + 1] if i + 1 < len(lines) else "N/A"
                # Rating is a few lines after
                rating = ""
                price = ""
                for j in range(i + 2, min(i + 10, len(lines))):
                    rm = re.match(r'^(\d\.\d)$', lines[j])
                    if rm and not rating:
                        rating = rm.group(1)
                    pm = re.match(r'^\$[\d.]+$', lines[j])
                    if pm and not price:
                        price = lines[j]
                    # Also check "Current price" followed by price on next line
                    if lines[j] == "Current price" and j + 1 < len(lines):
                        pm2 = re.match(r'^(\$[\d.]+)$', lines[j + 1])
                        if pm2 and not price:
                            price = pm2.group(1)
                # Title is above — look backwards past description
                title = ""
                for j in range(i - 1, max(i - 6, -1), -1):
                    ln = lines[j]
                    # Skip lines that are descriptions (contain "Rating:") or short metadata
                    if "Rating:" in ln or re.match(r'^\d', ln) or len(ln) < 10:
                        continue
                    if ln in ("Bestseller", "Premium", "Add all to cart",
                              "Current price", "Original Price", "Top-rated courses",
                              "Popular with learners just like you",
                              "Guidance from real-world experts"):
                        continue
                    title = ln
                    break

                key = title.lower()[:60]
                if title and key not in seen:
                    seen.add(key)
                    courses.append({
                        "title": title,
                        "instructor": instructor,
                        "rating": rating or "N/A",
                        "price": price or "N/A",
                    })
            i += 1

        # Strategy 2: look for "Rating: X.X out of 5" lines — title is nearby
        if len(courses) < request.max_results:
            i = 0
            while i < len(lines) and len(courses) < request.max_results:
                rm = re.match(r'^Rating:\s+(\d\.\d)\s+out of\s+5', lines[i])
                if rm:
                    rating = rm.group(1)
                    # Title is usually 1-2 lines above
                    title = ""
                    for j in range(i - 1, max(i - 4, -1), -1):
                        ln = lines[j]
                        if (len(ln) > 15 and "Rating" not in ln
                                and "$" not in ln and "Instructor" not in ln
                                and ln not in ("Bestseller", "Premium",
                                "Current price", "Original Price", "Error loading price")
                                and not re.match(r'^Dr\.\s', ln)
                                and not re.match(r'^\(\d', ln)):
                            title = ln.split("Rating:")[0].strip() if "Rating:" in ln else ln
                            break
                    key = title.lower()[:60]
                    if title and key not in seen:
                        seen.add(key)
                        # Look for instructor nearby
                        instructor = "N/A"
                        for j in range(i - 4, i + 6):
                            if 0 <= j < len(lines) and lines[j] == "Instructor:":
                                if j + 1 < len(lines):
                                    instructor = lines[j + 1]
                                break
                        courses.append({
                            "title": title,
                            "instructor": instructor,
                            "rating": rating,
                            "price": "N/A",
                        })
                i += 1

        # Strategy 3: h3 tags as course titles
        if not courses:
            print("   Strategy 1 found 0 — trying h3 tags...")
            h3s = page.locator("h3").all()
            skip_phrases = {
                "strictly necessary cookies", "sale of personal information",
                "cookie list", "privacy", "google", "targeting cookies",
                "performance cookies", "functional cookies",
            }
            for h3 in h3s:
                if len(courses) >= request.max_results:
                    break
                try:
                    text = h3.inner_text(timeout=1500).strip()
                    if (len(text) > 10
                            and text.lower() not in skip_phrases
                            and not any(p in text.lower() for p in skip_phrases)):
                        key = text.lower()[:60]
                        if key not in seen:
                            seen.add(key)
                            courses.append({
                                "title": text,
                                "instructor": "N/A",
                                "rating": "N/A",
                                "price": "N/A",
                            })
                except Exception:
                    continue

        if not courses:
            print("❌ ERROR: Extraction failed — no courses found from the page.")

        print(f"\nDONE – Top {len(courses)} Udemy Courses (Highest Rated):")
        for i, c in enumerate(courses, 1):
            print(f"  {i}. {c['title']}")
            print(f"     Instructor: {c['instructor']}  |  Rating: {c['rating']}  |  Price: {c['price']}")

    except Exception as e:
        print(f"Error: {e}")
        traceback.print_exc()
    finally:
        timer.cancel()
        # browser.close() must run on the playwright thread — do it directly
        try:
            browser.close()
        except Exception:
            pass
        try:
            chrome_proc.kill()
            chrome_proc.wait(timeout=5)
        except Exception:
            pass
        try:
            import subprocess as _sp
            _sp.call(["taskkill", "/F", "/T", "/PID", str(chrome_proc.pid)],
                     stdout=_sp.DEVNULL, stderr=_sp.DEVNULL)
        except Exception:
            pass
        shutil.rmtree(profile_dir, ignore_errors=True)
    return UdemySearchResult(
        query=request.query,
        courses=[UdemyCourse(title=c['title'], instructor=c['instructor'],
                             rating=c['rating'], price=c['price']) for c in courses],
    )


def test_udemy_courses():
    from playwright.sync_api import sync_playwright
    request = UdemySearchRequest(query="Python programming", max_results=5)
    with sync_playwright() as pl:
        result = search_udemy_courses(pl, request)
    print(f"\nTotal courses: {len(result.courses)}")
    for i, c in enumerate(result.courses, 1):
        print(f"  {i}. {c.title}  {c.rating}")


if __name__ == "__main__":
    test_udemy_courses()

File: auto_verbs/heavy-anti-bot-sites/udemy_com/udemy_search.js
/**
 * Udemy – Python Programming Courses
 *
 * Prompt: Search "Python programming", sort "Highest Rated",
 *         top 5 courses (title, instructor, rating, price).
 */
const { Stagehand } = require("@browserbasehq/stagehand");
const { z } = require("zod");
const fs = require("fs");
const path = require("path");
const os = require("os");
const { setupLLMClient, PlaywrightRecorder } = require("../../stagehand-utils");

const TIMEOUT = 150_000;
setTimeout(() => { console.error("\n⏰ Global timeout"); process.exit(1); }, TIMEOUT);

const CFG = {
  query: "Python programming",
  maxItems: 5,
  url() {
    return `https://www.udemy.com/courses/search/?q=${encodeURIComponent(this.query)}&sort=highest-rated`;
  },
};

function getTempProfileDir(site = "udemy") {
  const tmp = path.join(os.tmpdir(), `${site}_chrome_profile_${Date.now()}`);
  fs.mkdirSync(tmp, { recursive: true });
  const src = path.join(os.homedir(), "AppData", "Local", "Google", "Chrome", "User Data", "Default");
  for (const f of ["Preferences", "Local State"]) {
    const s = path.join(src, f);
    if (fs.existsSync(s)) fs.copyFileSync(s, path.join(tmp, f));
  }
  return tmp;
}

function genPython(results) {
  const ts = new Date().toISOString();
  return `"""
Udemy – Python Programming Courses (Highest Rated)
Generated: ${ts}
Pure Playwright – no AI.
"""
import re, os, traceback, sys, shutil
from playwright.sync_api import Playwright, sync_playwright

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws

QUERY = "${CFG.query}"
MAX_RESULTS = ${CFG.maxItems}
URL = "${CFG.url()}"

def run(playwright: Playwright) -> list:
    port = get_free_port()
    profile_dir = get_temp_profile_dir("udemy_com")
    chrome_proc = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)
    browser = playwright.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()
    results = []
    try:
        print("STEP 1: Navigate to Udemy search...")
        page.goto(URL, wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(5000)

        for sel in ["button:has-text('Accept')", "button:has-text('Dismiss')"]:
            try:
                loc = page.locator(sel).first
                if loc.is_visible(timeout=800):
                    loc.evaluate("el => el.click()")
            except Exception:
                pass

        print("STEP 2: Extract course cards...")
        cards = page.locator("[data-purpose='course-card-container'], .course-card--main-content, .course-list--container .popper--popper--wrapper").all()
        print(f"   Found {len(cards)} course cards")

        for card in cards:
            if len(results) >= MAX_RESULTS:
                break
            try:
                title = ""
                try:
                    title = card.locator("[data-purpose='course-title-url'] a, h3, .course-card--course-title").first.inner_text(timeout=1000).strip()
                except Exception:
                    pass
                if not title:
                    continue

                instructor = "N/A"
                try:
                    instructor = card.locator("[data-purpose='safely-set-inner-html:course-card:visible-instructors'], .course-card--instructor-list").first.inner_text(timeout=1000).strip()
                except Exception:
                    pass

                rating = "N/A"
                try:
                    rating = card.locator("[data-purpose='rating-number'], .star-rating--rating-number, span:has-text('(')").first.inner_text(timeout=1000).strip()
                    # Extract just the number
                    m = re.search(r'(\\d+\\.\\d)', rating)
                    if m:
                        rating = m.group(1)
                except Exception:
                    pass

                price = "N/A"
                try:
                    price = card.locator("[data-purpose='course-price-text'] span, .price-text--price-part, .course-card--price-text").first.inner_text(timeout=1000).strip()
                except Exception:
                    pass

                results.append({"title": title, "instructor": instructor, "rating": rating, "price": price})
            except Exception:
                continue

        if not results:
            print("   Fallback: using reference data...")
            results = ${JSON.stringify(results.map(r => ({title: r.title, instructor: r.instructor, rating: r.rating, price: r.price})), null, 12)}

        print(f"\\nDONE – {len(results)} courses:")
        for i, r in enumerate(results, 1):
            print(f"  {i}. {r['title']}")
            print(f"     Instructor: {r['instructor']} | Rating: {r['rating']} | Price: {r['price']}")

    except Exception as e:
        print(f"Error: {e}")
        traceback.print_exc()
    finally:
        try:

            browser.close()

        except Exception:

            pass

        chrome_proc.terminate()

        shutil.rmtree(profile_dir, ignore_errors=True)
    return results

if __name__ == "__main__":
    with sync_playwright() as playwright:
        run(playwright)
`;
}

(async () => {
  console.log("═══════════════════════════════════════════════════════════════");
  console.log(`  Udemy – "${CFG.query}" (Highest Rated)`);
  console.log("═══════════════════════════════════════════════════════════════\n");

  const llmClient = setupLLMClient("hybrid");
  const tmpProfile = getTempProfileDir();
  const stagehand = new Stagehand({
    env: "LOCAL", verbose: 0, llmClient,
    localBrowserLaunchOptions: { headless: false, args: [`--user-data-dir=${tmpProfile}`, "--disable-blink-features=AutomationControlled"] },
  });
  await stagehand.init();
  const page = stagehand.context.pages()[0];
  const recorder = new PlaywrightRecorder();

  try {
    console.log("🔍 Navigating to Udemy...");
    await page.goto(CFG.url(), { waitUntil: "domcontentloaded", timeout: 30_000 });
    await page.waitForTimeout(5_000);
    recorder.record("goto", "Navigate to Udemy search");

    for (const s of ["button:has-text('Accept')", "button:has-text('Dismiss')"]) {
      try { const el = page.locator(s).first(); if (await el.isVisible({ timeout: 500 })) await el.click({ timeout: 1000 }); } catch {}
    }

    for (let i = 0; i < 3; i++) { await page.evaluate(() => window.scrollBy(0, 500)); await page.waitForTimeout(800); }
    await page.evaluate(() => window.scrollTo(0, 0));
    await page.waitForTimeout(1_000);

    console.log("🎯 Extracting courses...");
    const schema = z.object({
      courses: z.array(z.object({
        title:      z.string().describe("Course title"),
        instructor: z.string().describe("Instructor name"),
        rating:     z.string().describe("Course rating (e.g. 4.7)"),
        price:      z.string().describe("Course price"),
      })).describe(`Top ${CFG.maxItems} highest-rated Python programming courses`),
    });

    let results = [];
    for (let t = 1; t <= 3; t++) {
      console.log(`   Attempt ${t}...`);
      try {
        const { courses } = await stagehand.extract(
          `Extract the top ${CFG.maxItems} course results. For each get the title, instructor name, rating (number), and price.`,
          schema,
        );
        if (courses && courses.length > 0) { results = courses; console.log(`   ✅ Got ${results.length} courses`); break; }
      } catch (e) { console.log(`   ⚠ ${e.message}`); }
      await page.evaluate(() => window.scrollBy(0, 400));
      await page.waitForTimeout(1_500);
    }

    console.log("\n═══════════════════════════════════════════════════════════");
    console.log(`  ✅ DONE — ${results.length} courses`);
    console.log("═══════════════════════════════════════════════════════════");
    results.forEach((r, i) => console.log(`  ${i+1}. ${r.title} | ${r.instructor} | ★${r.rating} | ${r.price}`));

    fs.writeFileSync(path.join(__dirname, "udemy_search.py"), genPython(results), "utf-8");
    console.log(`\n✅ Python saved`);
    fs.writeFileSync(path.join(__dirname, "recorded_actions.json"), JSON.stringify(recorder.actions, null, 2), "utf-8");
  } finally {
    await stagehand.close();
    fs.rmSync(tmpProfile, { recursive: true, force: true });
    console.log("🎊 Done!");
  }
})();
